### PR TITLE
[confmap] Set confmap.unifyEnvVarExpansion feature gate to stable

### PR DESCRIPTION
**Pull request by TylerHelmuth**
#### Description

This PR promotes the `confmap.unifyEnvVarExpansion` feature gate to stable and sets a `ToVersion` of `v0.106.0`, anticipating that the gate be completely removed in that version.

We should weigh if switching the Stable should be done in `v0.105.0` or if it needs more time in `Beta` to give users more time to switch.  Delaying promotion to `Stable` delays confmap 1.0.

If we merge this we need to commit to merging https://github.com/open-telemetry/opentelemetry-collector/pull/10510 in the same release.

#### Link to tracking issue
Related to https://github.com/open-telemetry/opentelemetry-collector/issues/10161
Related to https://github.com/open-telemetry/opentelemetry-collector/issues/7111
Related to https://github.com/open-telemetry/opentelemetry-collector/issues/8215

### CODE REVIEW

File: confmap/converter/expandconverter/expand_test.go
Comment (TylerHelmuth) [inline confmap/converter/expandconverter/expand_test.go:1]: Make the feature gate stable negates these tests, which set the gate to false.  Since the module is being deprecated, and will be removed shortly after, removing the tests doesn't feel like a big deal.

**Comment by TylerHelmuth** [comment]
Let me know if you'd rather this PR and https://github.com/open-telemetry/opentelemetry-collector/pull/10510 be merged into 1.  If we needed to roll back for some reason 1 PR would be easier.

**Comment by mx-psi** [comment]
I think we can wait at least until #10554 has been out for one release

**Comment by ben-childs-docusign** [comment]
Hi Folks - I just logged this issue it seems like there is still an issue with configurations that use "$$SomeString" syntax to represent the string literal "$SomeString" with this feature flag enabled it is still blocked on startup with version 105.

https://github.com/open-telemetry/opentelemetry-collector/issues/10713

**Comment by mx-psi** [comment]
@ben-childs-docusign Thanks for the report. Have you tried v0.106.1? The fix for that issue was released on v0.106.0

**Comment by ben-childs-docusign** [comment]
Hi @mx-psi yes we deployed the patch and it is working. thanks for the fix!

File: confmap/resolver.go
Comment (dmitryax) [inline confmap/resolver.go:174]: This can also be `expandedValue`, `[]any`, and `map[string]any`. We need to handle it a  function calling itself for each of the cases

Comment (TylerHelmuth) [reply confmap/resolver.go]: Updated to handle new types and added more test cases to the e2e expansion tests.

File: confmap/resolver.go
Comment (dmitryax) [inline confmap/resolver.go:203]: @mx-psi do we need this escaping here? Isn't the Original field supposed to keep the string as is?

**Review by mx-psi — APPROVED** [review]

File: confmap/resolver.go
Comment (mx-psi) [reply confmap/resolver.go:203]: @dmitryax Original is meant to keep the string representation. I think there are situations where an user would expect expansion to happen and situations where they wouldn't. 

#### Situation where an user expects expansion (and therefore escaping) to happen:

This is a bit of a contrived example but consider the following env vars and files:
- `URL_WITHOUT_PROTOCOL` set to `//example.com/api/v1/traces`
- `HTTP_URL` set to `http:${env:URL_WITHOUT_PROTOCOL}`
- a config file that sets the `otlp::endpoint` to `HTTP_URL`

In this situation, I would expect the OTLP endpoint to get `http://example.com/api/v1/traces`. 

Before #10618, this would fail with the `env` provider, since the contents of `HTTP_URL` would be parsed as a `map[string]any` with a single `http` key.

#### Situation where they do not expect expansion (or escaping) to happen

A password, that could have arbitrary contents

----

I don't know what to make out of this, we could have a third path so if a field is a `configopaque.String` we use the raw literal string (that's definitely doable with a bit of extra work and plumbing). I don't know if my imagination is good enough to cover all cases

Comment (TylerHelmuth) [reply confmap/resolver.go]: The tests still pass without the `OriginalValue` being escaped. Is there a scenario I could add to the e2e tests that would fail if `OriginalValue` isn't escape (or an inverse test if it should definitely not be escaped)

File: confmap/internal/e2e/expand_test.go
Comment (dmitryax) [inline confmap/internal/e2e/expand_test.go:23]: Given that $$ENV isn't supported anymore, it's better to add `$${}` here I think?
```suggestion
	t.Setenv("ENV_LIST", "['$$ESCAPE_ME','$${ESCAPE_ME}','${ENV_VALUE}']")
```

**Review by dmitryax — APPROVED** [review]

File: confmap/internal/e2e/expand_test.go
Comment (TylerHelmuth) [reply confmap/internal/e2e/expand_test.go]: Good idea, added

File: .chloggen/unify-env-var-gate-stable.yaml
Comment (evan-bradley) [inline .chloggen/unify-env-var-gate-stable.yaml:10]:
```suggestion
note: Set the `confmap.unifyEnvVarExpansion` feature gate to Stable. Expansion of `$FOO` env vars is no longer supported.  Use `${FOO}` or `${env:FOO}` instead.
```

File: confmap/converter/expandconverter/go.mod
Comment (evan-bradley) [inline confmap/converter/expandconverter/go.mod:6]:
```suggestion
	go.opentelemetry.io/collector/confmap v0.106.1
```

File: confmap/internal/e2e/expand_test.go
Comment (evan-bradley) [inline confmap/internal/e2e/expand_test.go:62]:
```suggestion
```
I think the function name should cover this, especially now that the expand converter is no longer involved.

File: confmap/internal/e2e/expand_test.go
Comment (evan-bradley) [inline confmap/internal/e2e/expand_test.go:19]:
```suggestion
```

**Review by evan-bradley — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]

**Comment by mx-psi** [comment]
I am going to wait to merge this until #10510 is approved

File: internal/globalgates/globalgates.go
Comment (TylerHelmuth) [inline internal/globalgates/globalgates.go:11]:
```suggestion
	featuregate.WithRegisterToVersion("v0.109.0"),
```

**Pull request merged**
